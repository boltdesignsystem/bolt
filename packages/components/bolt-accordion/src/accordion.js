import { unsafeCSS, BoltElement, customElement, html } from '@bolt/element';
import { withContext } from 'wc-context/lit-element';
import { smoothScroll } from '@bolt/components-smooth-scroll/src/smooth-scroll';
import URLSearchParams from '@ungap/url-search-params'; // URLSearchParams poly for older browsers
import classNames from 'classnames/bind';
import styles from './accordion.scss';
import schema from '../accordion.schema';
import { Accordion } from './_accordion-handorgel';

let cx = classNames.bind(styles);

@customElement('bolt-accordion')
class BoltAccordion extends withContext(BoltElement) {
  static schema = schema;

  static get properties() {
    return {
      ...this.props,
    };
  }

  static get providedContexts() {
    return {
      noSeparator: { property: 'noSeparator' },
      boxShadow: { property: 'boxShadow' },
      spacing: { property: 'spacing' },
      iconValign: { property: 'iconValign' },
    };
  }

  static get styles() {
    return [unsafeCSS(styles)];
  }

  updated(changedProperties) {
    super.updated && super.updated(changedProperties);
    let hasSpacingChanged = false;

    changedProperties.forEach((oldValue, propName) => {
      // is spacing has changed, wait for the updates to finish before updating handorgel
      if (propName === 'spacing') {
        hasSpacingChanged = true;
      }
    });

    this.updateComplete.then(() => {
      if (this.accordion && this.accordion.options) {
        this.accordion.options = this.accordionOptions;
        this.accordion.update();
      }

      if (hasSpacingChanged) {
        this.accordion?.resize();
        hasSpacingChanged = false;
      }
    });
  }

  get accordionOptions() {
    return {
      // array of node references, alternating trigger + content: [trigger, content, trigger, content]
      items: this.accordionItems,

      // whether multiple folds can be opened at once
      multiSelectable: !this.single,
      // whether the folds are collapsible
      collapsible: true,

      // whether ARIA attributes are enabled
      ariaEnabled: true,
      // whether W3C keyboard shortcuts are enabled
      keyboardInteraction: true,
      // whether to loop trigger focus (sets focus back to first/last trigger when end/start reached)
      carouselFocus: true,

      // attribute for the trigger or content to open folds at initialization
      initialOpenAttribute: 'data-open',
      // whether to use transition at initial open
      initialOpenTransition: true,
      // delay used to show initial transition
      initialOpenTransitionDelay: 200,

      // trigger/content class if fold is open
      triggerOpenClass: 'c-bolt-accordion-item__trigger--open',
      contentOpenClass: 'c-bolt-accordion-item__content--open',

      // trigger/content class if fold has been opened (transition finished)
      triggerOpenedClass: 'c-bolt-accordion-item__trigger--opened',
      contentOpenedClass: 'c-bolt-accordion-item__content--opened',

      // trigger/content class if fold has been focused
      triggerFocusClass: 'c-bolt-accordion-item__trigger--focus',
      contentFocusClass: 'c-bolt-accordion-item__content--focus',

      // trigger/content class if fold is disabled
      triggerDisabledClass: 'c-bolt-accordion-item__trigger--disabled',
      contentDisabledClass: 'c-bolt-accordion-item__content--disabled',

      // trigger/content class if no transition should be active (applied on resize)
      triggerNoTransitionClass: 'c-bolt-accordion-item__trigger--notransition',
      contentNoTransitionClass: 'c-bolt-accordion-item__content--notransition',
    };
  }

  connectedCallback() {
    super.connectedCallback && super.connectedCallback();

    const urlParams = new URLSearchParams(window.location.search);
    const selectedItemParam = urlParams.get('selected-accordion-item');
    this.deepLinkTarget = this.querySelector(
      `#${selectedItemParam}:not([inactive])`,
    );

    if (this.single && this.deepLinkTarget) {
      // When in "single" mode, a deep link should override any items set to auto-open.
      // Unset these items immediately or we face race conditions as Handorgel initializes.
      this.querySelectorAll('bolt-accordion-item[open]').forEach(el =>
        el.removeAttribute('open'),
      );
    }

    this.addEventListener('bolt:layout-size-changed', this.handleLayoutChanged);
  }

  handleAccordionItemReady(item) {
    const root = item.renderRoot;
    const trigger =
      root && root.querySelector('.c-bolt-accordion-item__trigger');
    const content =
      root && root.querySelector('.c-bolt-accordion-item__content');

    if (trigger && content) {
      this.accordionItems = [...this.accordionItems, trigger, content];
    }

    const index = this.accordionItemElementsCopy.indexOf(item);

    if (index !== -1) {
      this.accordionItemElementsCopy.splice(index, 1);
    }

    if (this.accordionItemElementsCopy.length === 0) {
      if (this.accordion) {
        this.accordion.update();
      } else {
        this.initAccordion();
      }
    }
  }

  initAccordion() {
    this.accordion = new Accordion(
      this.accordionElement,
      this.accordionOptions,
    );

    this.accordion.folds.forEach(fold => {
      if (
        fold.button.classList.contains(
          'c-bolt-accordion-item__trigger-label--inactive',
        )
      ) {
        fold.disable();
      }
    });

    this.accordion.on('destroyed', fold => {
      delete this.accordion;
    });

    this.accordion.on('fold:opened', fold => {
      this.dispatchLayoutChanged();

      // @todo: register these elements in Bolt data instead?
      const elementsToUpdate = this.querySelectorAll('[will-update]');
      if (elementsToUpdate.length) {
        elementsToUpdate.forEach(el => {
          el.updateLayout && el.updateLayout();
        });
      }
    });

    this.accordion.on('fold:closed', fold => {
      this.dispatchLayoutChanged();
    });
  }

  dispatchLayoutChanged() {
    this.dispatchEvent(
      new CustomEvent('bolt:layout-size-changed', {
        bubbles: true,
      }),
    );
  }

  handleLayoutChanged(e) {
    if (e.target !== this) {
      this.accordion && this.accordion.resize();
    }
  }

  setupAccordion() {
    // Accordion outer container, accordion plugin will be instantiated on this element
    this.accordionElement = this.renderRoot.querySelector('.c-bolt-accordion');

    // Root used to find child accordion items
    this.accordionRoot = this.useShadow ? this : this.accordionElement;

    // Reference to all the accordion items
    this.accordionItemElements = Array.from(this.accordionRoot.children).filter(
      item => item.tagName === 'BOLT-ACCORDION-ITEM',
    );

    // Copy of that reference that will be mutated as we process each accordion item
    this.accordionItemElementsCopy = [...this.accordionItemElements];

    // Array passed to the Accordion plugin, a series of trigger/content pairs
    this.accordionItems = [];

    Promise.all(
      this.accordionItemElements.map(item => {
        if (item._wasInitiallyRendered || this._wasMutated) return;
        return new Promise((resolve, reject) => {
          item.addEventListener('ready', e => {
            return item === e.target && resolve();
          });
          item.addEventListener('error', reject);
        });
      }),
    ).then(() => {
      this.accordionItemElements.forEach(item => {
        this.handleAccordionItemReady(item);
      });
      this.handleDeepLink();
      this._wasMutated = false;
    });

    setTimeout(() => {
      // "Open item" hack: This hack is required to fix accordion items that are supposed to open on load but do not because they are inside a hidden element, e.g. an inactive tab panel
      const openItems = this.querySelectorAll('bolt-accordion-item[open]');
      if (openItems.length && openItems[0].offsetWidth === 0) {
        this.openItemToFix = openItems[0];
        this.setAttribute('will-update', '');
      }
    }, 250); // timeout allows parent components (such as Tabs) to finish rendering
  }

  // Public method called when a Tab Panel that contains an Accordion is opened and that Accordion has been flagged with the "[will-update]" attribute
  updateLayout() {
    // "Open item" hack continued...
    if (this.openItemToFix) {
      const index = Array.from(this.accordionItemElements).indexOf(
        this.openItemToFix,
      );
      this.accordion.folds[index].close();
      this.accordion.folds[index].open();
      this.openItemToFix = null;
      this.removeAttribute('will-update', '');
    }
  }

  addMutationObserver() {
    const self = this;

    // todo: this.useShadow is a temporary workaround until mutation observer works better with light DOM
    if (window.MutationObserver && this.useShadow) {
      // Re-generate slots + re-render when mutations are observed
      const mutationCallback = function(mutationsList, observer) {
        for (let mutation of mutationsList) {
          if (mutation.type === 'childList') {
            // @todo: handle add/remove children
            // console.log('A child node has been added or removed.');
            // self.slots = self._checkSlots();
            // self.triggerUpdate();
          } else if (mutation.type === 'attributes') {
            if (mutation.target.tagName === 'BOLT-ACCORDION-ITEM') {
              const target = mutation.target;

              this._wasMutated = true;

              const resetAccordion = () => {
                self.accordion.destroy();
                self.setupAccordion();
                target.removeEventListener('rendered', resetAccordion);
              };

              target.addEventListener('rendered', resetAccordion);
            }
          }
        }
      };

      // Create an observer instance linked to the callback function
      self.observer = new MutationObserver(mutationCallback);

      // Start observing the target node for configured mutations
      self.observer.observe(this, {
        attributes: true,
        childList: true,
        subtree: true,
      });
    }
  }

  async handleDeepLink() {
    if (!this.deepLinkTarget) return;

    const deepLinkTargetIndex = this.accordionItemElements.indexOf(
      this.deepLinkTarget,
    );
    let shouldScrollIntoView;

    if (deepLinkTargetIndex !== -1) {
      // This Promise is a workaround for a bug that sometimes happens when you
      // put an accordion in a band and use deep linking. When the band first
      // renders, it interrupts the initial accordion animation and causes
      // Handorgel's 'transitionend' event to not fire. @see DS-253 for more.
      const closestBand = this.closest('bolt-band');
      if (closestBand) {
        await new Promise((resolve, reject) => {
          closestBand._wasInitiallyRendered && resolve();
          closestBand.addEventListener('ready', e => {
            e.target === closestBand && resolve();
          });
          closestBand.addEventListener('error', reject);
        });
      }

      this.accordion.folds[deepLinkTargetIndex].open();
      shouldScrollIntoView = true;
    }

    if (shouldScrollIntoView) {
      let shouldResetScroll;

      if (window.history?.scrollRestoration === 'auto') {
        // If you are refreshing the page and using a browser with `scrollRestoration`,
        // temporarily disable `scrollRestoration` while we scroll to the element, avoids janky scroll.
        // https://developers.google.com/web/updates/2015/09/history-api-scroll-restoration
        window.history.scrollRestoration = 'manual';
        shouldResetScroll = true;
      }

      setTimeout(() => {
        smoothScroll.animateScroll(this.deepLinkTarget, 0, {
          header: this.scrollOffsetSelector,
          offset: this.scrollOffset || 0,
          speed: 750,
          easing: 'easeInOutCubic',
          updateURL: false,
        });

        shouldScrollIntoView = false;

        if (shouldResetScroll) {
          setTimeout(() => {
            window.history.scrollRestoration = 'auto';
          }, 1000); // wait another second to turn 'scrollRestoration' back on, just to be safe
        }
      }, 750); // Must let the page load or scroll is not at all "smooth", can reduce to 500ms but not much less
    }
  }

  firstUpdated(changedProperties) {
    super.firstUpdated && super.firstUpdated(changedProperties);

    if (this.accordion) {
      // If accordion already exists, update options
      this.accordion.options = this.accordionOptions;
    } else {
      this.setupAccordion();
    }
    if (!this.observer) {
      this.addMutationObserver();
    }
  }

  disconnectedCallback() {
    super.disconnectedCallback && super.disconnectedCallback();

    this.removeEventListener(
      'bolt:layout-size-changed',
      this.handleLayoutChanged,
    );

    // remove MutationObserver if supported + exists
    if (window.MutationObserver && this.observer) {
      this.observer.disconnect();
    }
  }

  render() {
    const classes = cx('c-bolt-accordion', {
      'c-bolt-accordion--box-shadow': this.boxShadow,
    });

    return html`
      <div class="${classes}">
        ${this.slotify('default')}
      </div>
    `;
  }
}

export { BoltAccordion };
