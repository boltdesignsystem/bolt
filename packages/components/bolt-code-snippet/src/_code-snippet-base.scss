@import '@bolt/core-v3.x';

/* ------------------------------------ *\
    Bolt Code Snippet
\* ------------------------------------ */

.c-bolt-code-snippet {
  overflow: hidden;
  margin: 0 0 var(--bolt-spacing-y-medium) 0;
  border-radius: bolt-border-radius(small);

  &:last-child {
    margin-bottom: 0;
  }

  pre {
    max-height: 50vh;
    border: 0;
    box-shadow: none;
  }
}

.c-bolt-code-snippet__header,
.c-bolt-code-snippet:before {
  padding: calc(var(--bolt-spacing-y-small) / 2) var(--bolt-spacing-x-small);
  font-family: var(--bolt-type-font-family-headline);
  font-size: var(--bolt-type-font-size-xsmall);
  font-weight: var(--bolt-type-font-weight-semibold);
  color: var(--bolt-color-gray-dark);
  line-height: var(--bolt-type-line-height-xsmall);
  border-bottom: 1px solid var(--bolt-color-white);
  background: inherit;
}

.c-bolt-code-snippet__header {
  display: flex;
  justify-content: space-between;
}

.c-bolt-code-snippet--dark {
  &:before {
    border-bottom-color: var(--bolt-color-gray-dark);
  }

  .c-bolt-code-snippet__header {
    color: var(--bolt-color-gray);
    border-bottom-color: var(--bolt-color-gray-dark);
  }
}

.c-bolt-code-snippet:not([data-bolt-hide-lang-label][data-bolt-hide-copy]):not([data-bolt-ready]):before {
  content: '\00a0'; // Non-breaking space provides same height as header (with text) once JS loads, preventing layout shift.
  display: block;
}

.c-bolt-code-snippet__copy {
  margin-left: auto;

  .e-bolt-text-link:focus {
    box-shadow: 0 0 0 1px var(--bolt-color-navy-light);
    outline: none; // Fix a Safari bug which the focus outline does not appear.
  }
}
