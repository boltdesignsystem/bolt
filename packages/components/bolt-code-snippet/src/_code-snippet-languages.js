/*
Language map, converts language id to readable title.
*/

export const languages = {
  // Base languages...
  markup: 'Markup',

  // "Markup" includes:
  html: 'HTML',
  xml: 'XML',
  svg: 'SVG',
  mathml: 'MathML',
  ssml: 'SSML',
  atom: 'Atom',
  rss: 'RSS',

  // Additional base languages:
  javascript: 'JavaScript',
  js: 'JavaScript',
  clike: 'C-like',
  css: 'CSS',

  // Begin Add-on languages...
  scss: 'Sass (Scss)',
  twig: 'Twig',
  java: 'Java',
  json: 'JSON',
  rest: 'reST (reStructuredText)',
  bash: 'Bash',
  shell: 'Bash',
  csv: 'CSV',
  docker: 'Docker',
  dockerfile: 'Docker',
  http: 'HTTP',
  jsx: 'React JSX',
  tsx: 'React TSX',
  md: 'Markdown',
  markdown: 'Markdown',
  yml: 'YAML',
  yaml: 'YAML',
};
