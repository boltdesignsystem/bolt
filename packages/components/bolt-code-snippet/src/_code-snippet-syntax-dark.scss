// Dark Syntax

.c-bolt-code-snippet--dark {
  &,
  code,
  pre {
    color: var(--bolt-color-gray-xlight);
    background: var(--bolt-color-navy-xdark);
  }

  pre::-moz-selection,
  pre ::-moz-selection,
  code::-moz-selection,
  code ::-moz-selection {
    background: var(--bolt-color-navy-xlight);
  }

  pre::selection,
  pre ::selection,
  code::selection,
  code ::selection {
    background: var(--bolt-color-navy-xlight);
  }

  .token.comment,
  .token.prolog,
  .token.doctype,
  .token.cdata {
    color: #5c6370;
  }

  .token.punctuation {
    color: #abb2bf;
  }

  .token.selector,
  .token.tag {
    color: lightsalmon;
  }

  .token.property,
  .token.boolean,
  .token.number,
  .token.constant,
  .token.symbol,
  .token.attr-name,
  .token.deleted {
    color: #d19a66;
  }

  .token.string,
  .token.char,
  .token.attr-value,
  .token.builtin,
  .token.inserted {
    color: #98c379;
  }

  .token.operator,
  .token.entity,
  .token.url,
  .language-css .token.string,
  .style .token.string {
    color: #56b6c2;
  }

  .token.atrule,
  .token.keyword {
    color: #c678dd;
  }

  .token.function {
    color: #61afef;
  }

  .token.regex,
  .token.important,
  .token.variable {
    color: #c678dd;
  }

  .line-numbers-rows > span:before {
    color: #5c6370;
  }
}
