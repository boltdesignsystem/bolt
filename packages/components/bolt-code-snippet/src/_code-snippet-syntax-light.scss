// Light Syntax

.c-bolt-code-snippet--light {
  // Redefine vars that text link styles are dependent on.
  --m-bolt-text: var(--bolt-color-black);
  --m-bolt-headline: var(--bolt-color-navy);
  --m-bolt-link: var(--bolt-color-navy-light);

  border: 1px solid var(--bolt-color-white);
  box-shadow: 0 0 6px rgba(bolt-color(gray, light), 0.8); // Sass function is used due to transparency needed.

  &,
  code,
  pre {
    color: var(--bolt-color-black);
    background: var(--bolt-color-gray-xlight);
  }

  pre::-moz-selection,
  pre ::-moz-selection,
  code::-moz-selection,
  code ::-moz-selection {
    background: var(--bolt-color-navy-xlight);
  }

  pre::selection,
  pre ::selection,
  code::selection,
  code ::selection {
    background: var(--bolt-color-navy-xlight);
  }

  .token.comment,
  .token.prolog,
  .token.doctype,
  .token.cdata {
    color: slategray;
  }

  .token.punctuation {
    color: #999;
  }

  .token.property,
  .token.tag,
  .token.boolean,
  .token.number,
  .token.constant,
  .token.symbol,
  .token.deleted {
    color: #905;
  }

  .token.selector,
  .token.attr-name,
  .token.string,
  .token.char,
  .token.builtin,
  .token.inserted {
    color: green;
  }

  .token.operator,
  .token.entity,
  .token.url,
  .language-css .token.string,
  .style .token.string {
    color: #9a6e3a;
    background: hsla(0, 0%, 100%, 0.5);
  }

  .token.atrule,
  .token.attr-value,
  .token.keyword {
    color: #07a;
  }

  .token.function,
  .token.class-name {
    color: #dd4a68;
  }

  .token.regex,
  .token.important,
  .token.variable {
    color: #e90;
  }

  .line-numbers-rows > span:before {
    color: slategray;
  }
}
