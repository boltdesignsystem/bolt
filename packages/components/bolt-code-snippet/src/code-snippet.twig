{% set schema = bolt.data.components["@bolt-components-code-snippet"].schema %}

{% if enable_json_schema_validation %}
  {{ validate_data_schema(schema, _self) | raw }}
{% endif %}

{# Variables #}
{% set this = init(schema) %}
{% set attributes = create_attribute(attributes|default({})) %}

{# Array of classes based on the defined + default props #}
{% set classes = [
  'c-bolt-code-snippet',
  this.data.syntax.value != 'none' ? 'c-bolt-code-snippet--'  ~ this.data.syntax.value : '',
] %}

{% set attributes = attributes.setAttribute('data-bolt-lang', this.data.lang.value) %}

{% if this.data.hide_copy.value %}
  {% set attributes = attributes.setAttribute('data-bolt-hide-copy', '') %}
{% endif %}

{% if this.data.hide_title.value %}
  {% set attributes = attributes.setAttribute('data-bolt-hide-title', '') %}
{% endif %}

{% if lang_title %}
  {% set attributes = attributes.setAttribute('data-bolt-lang-title', lang_title) %}
{% endif %}

{# Template #}
<div {{ attributes.addClass(classes)}}>
<pre><code>{% spaceless %}{{ content|e }}{% endspaceless %}</code></pre>
</div>
