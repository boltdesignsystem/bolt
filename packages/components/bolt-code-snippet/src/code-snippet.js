import Prism from 'prismjs';
import { html, render, ifDefined } from '@bolt/element';
import cx from 'classnames/bind';
import ClipboardJS from 'clipboard';

// Prism only runs automatically on elements with the "language-*" class, but switch to manual just in case.
Prism.manual = true;

// Included by default: markup, html, xml, svg, mathml, ssml, atom, rss, javascript, js, clike
// Add-on languages:
import 'prismjs/components/prism-java';
import 'prismjs/components/prism-json';
import 'prismjs/components/prism-rest';
import 'prismjs/components/prism-bash';
import 'prismjs/components/prism-csv';
import 'prismjs/components/prism-docker';
import 'prismjs/components/prism-http';
import 'prismjs/components/prism-jsx';
import 'prismjs/components/prism-tsx';
import 'prismjs/components/prism-sql';
import 'prismjs/components/prism-css-extras';
import 'prismjs/components/prism-scss';
import 'prismjs/components/prism-twig';
import 'prismjs/components/prism-markdown';
import 'prismjs/components/prism-yaml';

// Note: If adding a new language, be sure to add corresponding label to `./_code-snippet-languages`
import { languages } from './_code-snippet-languages';

export class BoltCodeSnippet {
  constructor(el) {
    if (!el) return;
    this.el = el;
    this.init();
  }

  init() {
    this.preElement = this.el.querySelector('pre');
    this.codeElement = this.el.querySelector('code');
    this.lang =
      this.el.dataset.lang && this.el.dataset.lang !== 'none'
        ? this.el.dataset.lang
        : null;
    this.customLangLabel = this.el.dataset.customLangLabel;
    this.hideLangLabel =
      this.el.hasAttribute('data-hide-lang-label') ||
      (!this.lang && !this.customLangLabel);
    this.hideCopy = this.el.hasAttribute('data-hide-copy');

    if (!(this.preElement && this.codeElement)) return;

    this.originalHTML = this.codeElement.innerHTML;
    this.filteredHTML = this.replaceEntities(this.originalHTML);

    if (!(this.hideLangLabel && this.hideCopy)) {
      this.setupHeader();
    }

    if (this.lang) {
      this.highlightHTML();
    }

    this.el.setAttribute('data-bolt-ready', '');
  }

  replaceEntities(string) {
    // Twig encodes HTML entities. Replace them before syntax highlighting.
    return string
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>');
  }

  highlightHTML() {
    try {
      this.codeElement.innerHTML = Prism.highlight(
        this.filteredHTML,
        Prism.languages[this.lang],
      );
    } catch (error) {
      console.log(error);
    }
  }

  setupHeader() {
    this.header = document.createElement('div');
    this.header.classList.add('c-bolt-code-snippet__header');

    this.el.prepend(this.header);

    // Custom language label overrides default
    this.langLabel = this.customLangLabel || languages[this.lang];
    this.copied = false;

    this.renderHeader();

    if (!this.hideCopy) {
      // Must come after renderHeader
      this.setupClipboard();
    }
  }

  setupClipboard() {
    const clipboardTrigger = this.header.querySelector(
      '.js-bolt-code-snippet-copy-trigger',
    );

    if (!clipboardTrigger) return;

    this.codeElement?.setAttribute('tabindex', '-1');

    const clip = new ClipboardJS(clipboardTrigger, {
      text: () => {
        return this.filteredHTML;
      },
    });

    clip.on('success', () => {
      this.copied = true;

      this.renderHeader();
      this.codeElement?.focus();

      setTimeout(() => {
        this.copied = false;
        this.renderHeader();
      }, 2000);
    });
    clip.on('error', function() {
      console.log('There was an error copying this code');
      this.copied = false;
    });
  }

  renderHeader() {
    const copyButtonClasses = cx(
      'e-bolt-text-link',
      'e-bolt-text-link--reversed-underline',
      'js-bolt-code-snippet-copy-trigger',
    );

    render(
      html`
        ${!this.hideLangLabel
          ? html`
              <span class="c-bolt-code-snippet__lang-label"
                >${this.langLabel}</span
              >
            `
          : ``}
        ${!this.hideCopy
          ? html`
              <div class="c-bolt-code-snippet__copy">
                <button
                  type="button"
                  class="${copyButtonClasses}"
                  hidden=${ifDefined(this.copied || undefined)}
                >
                  <span class="e-bolt-text-link__icon-before"
                    ><bolt-icon name="copy-to-clipboard"></bolt-icon></span
                  >Copy</button
                >${this.copied
                  ? html`
                      <span aria-live="assertive">
                        <span aria-hidden="true">Copied!</span>
                        <span class="u-bolt-visuallyhidden"
                          >Text is copied to clipboard.</span
                        >
                      </span>
                    `
                  : ``}
              </div>
            `
          : ``}
      `,
      this.header,
    );
  }
}
