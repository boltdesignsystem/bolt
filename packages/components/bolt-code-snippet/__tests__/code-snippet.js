// Refer to https://github.com/boltdesignsystem/bolt/wiki/Jest-Test:-Example-Jest-Test for more testing examples
import { render, stopServer } from '../../../testing/testing-helpers';
import schema from '../code-snippet.schema';
import { languages } from '../src/_code-snippet-languages';

const { lang, syntax } = schema.properties;
let page, fixtures;

afterAll(async () => {
  await stopServer();
  await page.close();
});

beforeEach(async () => {
  await page.evaluate(() => {
    document.body.innerHTML = '';
  });
});

beforeAll(async () => {
  page = await global.__BROWSER__.newPage();
  await page.goto('http://127.0.0.1:4444/', {
    timeout: 0,
  });

  const cssCode = `
    .my-css {
      display: block;
    }
  `;

  const jsCode = `import { props } from '@bolt/core-v3.x/utils';`;

  const twigCode = `
    {% include '@bolt-components-code-snippet/code-snippet.twig' with {
      lang: 'html',
      content: '<p>Hello world!</p>'
    } only %}
  `;

  const defaultData = {
    content: cssCode,
    lang: 'css',
  };

  const jsData = {
    content: jsCode,
    lang: 'js',
  };

  const twigData = {
    content: twigCode,
    lang: 'twig',
  };

  fixtures = {
    defaultData,
    jsData,
    twigData,
  };
});

describe('Bolt Code Snippet', () => {
  test('default', async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.defaultData,
      },
    );

    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });
  test('default with JS code', async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.jsData,
      },
    );

    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });
  test('default with Twig code', async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.twigData,
      },
    );

    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });
});

describe('Bolt Code Snippet Props', () => {
  // Target each of the schema keys with the following pattern
  syntax.enum.forEach(async option => {
    test(`syntax: ${option}`, async () => {
      const results = await render(
        '@bolt-components-code-snippet/code-snippet.twig',
        {
          ...fixtures.defaultData,
          syntax: option,
        },
      );

      await expect(results.ok).toBe(true);
      await expect(results.html).toMatchSnapshot();
    });
  });

  lang.enum.forEach(async option => {
    test(`lang: ${option}`, async () => {
      // Tests that every lang option has a corresponding title in the languages file
      const title = languages[option];
      await expect(title).toMatchSnapshot();
    });
  });

  test(`hide_title`, async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.defaultData,
        hide_title: true,
      },
    );

    await expect(results.ok).toBe(true);
    await expect(results.html).toMatchSnapshot();
  });

  test(`hide_copy`, async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.defaultData,
        hide_copy: true,
      },
    );

    await expect(results.ok).toBe(true);
    await expect(results.html).toMatchSnapshot();
  });

  test(`hide_title and hide_copy`, async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.defaultData,
        hide_title: true,
        hide_copy: true,
      },
    );

    await expect(results.ok).toBe(true);
    await expect(results.html).toMatchSnapshot();
  });

  test(`lang_title`, async () => {
    const results = await render(
      '@bolt-components-code-snippet/code-snippet.twig',
      {
        ...fixtures.defaultData,
        lang_title: 'Foo',
      },
    );

    await expect(results.ok).toBe(true);
    await expect(results.html).toMatchSnapshot();
  });
});
