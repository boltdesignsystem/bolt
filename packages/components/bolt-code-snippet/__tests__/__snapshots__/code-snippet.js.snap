// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bolt Code Snippet Props custom_lang_label 1`] = `
<div data-lang="css"
     data-mode="light"
     data-custom-lang-label="Foo"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props hide_copy 1`] = `
<div data-lang="css"
     data-mode="light"
     data-hide-copy
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props hide_lang_label 1`] = `
<div data-lang="css"
     data-mode="light"
     data-hide-lang-label
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props hide_lang_label and hide_copy 1`] = `
<div data-lang="css"
     data-mode="light"
     data-hide-copy
     data-hide-lang-label
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props lang: atom 1`] = `"Atom"`;

exports[`Bolt Code Snippet Props lang: bash 1`] = `"Bash"`;

exports[`Bolt Code Snippet Props lang: clike 1`] = `"C-like"`;

exports[`Bolt Code Snippet Props lang: css 1`] = `"CSS"`;

exports[`Bolt Code Snippet Props lang: csv 1`] = `"CSV"`;

exports[`Bolt Code Snippet Props lang: docker 1`] = `"Docker"`;

exports[`Bolt Code Snippet Props lang: dockerfile 1`] = `"Docker"`;

exports[`Bolt Code Snippet Props lang: html 1`] = `"HTML"`;

exports[`Bolt Code Snippet Props lang: http 1`] = `"HTTP"`;

exports[`Bolt Code Snippet Props lang: java 1`] = `"Java"`;

exports[`Bolt Code Snippet Props lang: javascript 1`] = `"JavaScript"`;

exports[`Bolt Code Snippet Props lang: js 1`] = `"JavaScript"`;

exports[`Bolt Code Snippet Props lang: json 1`] = `"JSON"`;

exports[`Bolt Code Snippet Props lang: jsx 1`] = `"React JSX"`;

exports[`Bolt Code Snippet Props lang: markdown 1`] = `"Markdown"`;

exports[`Bolt Code Snippet Props lang: markup 1`] = `"Markup"`;

exports[`Bolt Code Snippet Props lang: mathml 1`] = `"MathML"`;

exports[`Bolt Code Snippet Props lang: md 1`] = `"Markdown"`;

exports[`Bolt Code Snippet Props lang: none 1`] = `undefined`;

exports[`Bolt Code Snippet Props lang: rest 1`] = `"reST (reStructuredText)"`;

exports[`Bolt Code Snippet Props lang: rss 1`] = `"RSS"`;

exports[`Bolt Code Snippet Props lang: scss 1`] = `"Sass (Scss)"`;

exports[`Bolt Code Snippet Props lang: shell 1`] = `"Bash"`;

exports[`Bolt Code Snippet Props lang: ssml 1`] = `"SSML"`;

exports[`Bolt Code Snippet Props lang: svg 1`] = `"SVG"`;

exports[`Bolt Code Snippet Props lang: tsx 1`] = `"React TSX"`;

exports[`Bolt Code Snippet Props lang: twig 1`] = `"Twig"`;

exports[`Bolt Code Snippet Props lang: xml 1`] = `"XML"`;

exports[`Bolt Code Snippet Props lang: yaml 1`] = `"YAML"`;

exports[`Bolt Code Snippet Props lang: yml 1`] = `"YAML"`;

exports[`Bolt Code Snippet Props mode: dark 1`] = `
<div data-lang="css"
     data-mode="dark"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props mode: light 1`] = `
<div data-lang="css"
     data-mode="light"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet Props mode: none 1`] = `
<div data-lang="css"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet default 1`] = `
<div data-lang="css"
     data-mode="light"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      .my-css {
      display: block;
    }
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet default with JS code 1`] = `
<div data-lang="js"
     data-mode="light"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      import { props } from &#039;@bolt/core-v3.x/utils&#039;;
    </code>
  </pre>
</div>
`;

exports[`Bolt Code Snippet default with Twig code 1`] = `
<div data-lang="twig"
     data-mode="light"
     class="c-bolt-code-snippet"
>
  <pre>
    <code>
      {% include &#039;@bolt-components-code-snippet/code-snippet.twig&#039; with {
      lang: &#039;html&#039;,
      content: &#039;&lt;p&gt;Hello world!&lt;/p&gt;&#039;
    } only %}
    </code>
  </pre>
</div>
`;
